### PR TITLE
CIVDT-1273: Duplicate Deferment Account Numbers

### PR DESCRIPTION
**Pull request by dannybassi**

### CODE REVIEW

File: app/controllers/RepresentativeDanImportVATController.scala [hunk 1]
```diff
@@ -59,38 +60,50 @@ class RepresentativeDanImportVATController @Inject()(identify: IdentifierAction,
...
-          val authorityFiles = request.userAnswers.get(UploadAuthorityPage).getOrElse(Seq.empty).filterNot(_.dutyType == Vat)
-
-          for {
-            updatedAnswers <- Future.fromTry(request.userAnswers.set(UploadAuthorityPage, authorityFiles))
-            updatedAnswers <- Future.fromTry(updatedAnswers.set(CheckModePage, false))
-            updatedAnswers <- Future.fromTry(updatedAnswers.set(AdditionalDefermentTypePage, dan.danType))
-            updatedAnswers <- Future.fromTry(updatedAnswers.set(AdditionalDefermentNumberPage, dan.accountNumber))
-            _ <- sessionRepository.set(updatedAnswers)
+        val dutyAccountNumber = request.userAnswers.get(DefermentAccountPage).getOrElse("No Duty Account Number")
+        if (sameAccountNumber(dutyAccountNumber, dan.accountNumber, request.userAnswers)) {
```
Comment (jacemale-hmrc) [inline app/controllers/RepresentativeDanImportVATController.scala:64]: I'd probably avoid using `getOrElse` for this,
if we just want to compare the 2 numbers we can do it like
```suggestion
        val dutyAccountNumber = request.userAnswers.get(DefermentAccountPage)
        if (dutyAccountNumber.exists(sameAccountNumber(_, dan.accountNumber, request.userAnswers))) {
```

File: app/controllers/RepresentativeDanImportVATController.scala [hunk 2]
```diff
@@ -105,4 +118,13 @@ class RepresentativeDanImportVATController @Inject()(identify: IdentifierAction,
     }
   }
 
+  private[controllers] def sameAccountNumber(dutyAccountNumber: String,
+                                             vatAccountNumber: String,
+                                             userAnswers: UserAnswers) = {
+    lazy val dutyAccountNumber: String = userAnswers.get(DefermentAccountPage).getOrElse("")
```
Comment (jacemale-hmrc) [inline app/controllers/RepresentativeDanImportVATController.scala:124]: it doesn't need to be lazy
```suggestion
    val dutyAccountNumber: String = userAnswers.get(DefermentAccountPage).getOrElse("")
```

File: app/controllers/RepresentativeDanImportVATController.scala [hunk 3]
```diff
@@ -105,4 +118,13 @@ class RepresentativeDanImportVATController @Inject()(identify: IdentifierAction,
...
+    dutyAccountNumber.exists(
+      accountNumber => dutyAccountNumber == vatAccountNumber
+    )
```
Comment (jacemale-hmrc) [inline app/controllers/RepresentativeDanImportVATController.scala:127]: we don't need `getOrElse` here either`
```suggestion
    val dutyAccountNumber = userAnswers.get(DefermentAccountPage)
    dutyAccountNumber.exists(_ == vatAccountNumber)
```

File: test/controllers/RepresentativeDanImportVATControllerSpec.scala [hunk 4]
```diff
@@ -257,4 +257,36 @@ class RepresentativeDanImportVATControllerSpec extends ControllerSpecBase {
...
+        UserAnswers("some-cred-id")
+          .set(DefermentAccountPage, "1234567").success.value
+          .set(DefermentTypePage, "C").success.value
+          .set(AdditionalDefermentNumberPage, "1234567").success.value
+          .set(AdditionalDefermentTypePage, "A").success.value
+
+      )
+      controller.sameAccountNumber("1234567", "1234567", userAnswers.get) mustBe true
+      val request = fakeRequest.withFormUrlEncodedBody(buildForm(accountNumber = Some("1234567"), danType = Some("C")): _*)
+      controller.onSubmit(request)
```
Comment (jacemale-hmrc) [inline test/controllers/RepresentativeDanImportVATControllerSpec.scala]: this tests seems to submit a request but doesn't verify anything about it (status code or errors etc.)

**Review by jacemale-hmrc — APPROVED** [review]

File: app/controllers/RepresentativeDanImportVATController.scala [hunk 5]
```diff
@@ -59,38 +60,46 @@ class RepresentativeDanImportVATController @Inject()(identify: IdentifierAction,
...
+              updatedAnswers <- Future.fromTry(updatedAnswers.set(AdditionalDefermentNumberPage, dan.accountNumber))
+              _ <- sessionRepository.set(updatedAnswers)
+            } yield {
+              if (request.checkMode) {
+                Redirect(controllers.cya.routes.CheckYourAnswersController.onLoad())
+              } else {
+                dan.danType match {
+                  case "A" | "C" => Redirect(controllers.cya.routes.CheckYourAnswersController.onLoad())
+                  case _ => Redirect(controllers.routes.UploadAuthorityController.onLoad(Vat, dan.accountNumber))
+                }
```
Comment (tvaidotas) [inline app/controllers/RepresentativeDanImportVATController.scala:100]: duplicate, could be refactored into redirect method

File: app/controllers/RepresentativeDanImportVATController.scala [hunk 6]
```diff
@@ -59,38 +60,46 @@ class RepresentativeDanImportVATController @Inject()(identify: IdentifierAction,
...
+        if (dutyAccountNumber.contains(dan.accountNumber)) {
+          val form = formProvider().fill(RepresentativeDan(dan.accountNumber, dan.danType))
+            .withError(FormError("accountNumber", "repDan.error.input.sameAccountNumber"))
+          Future.successful(Ok(view(form, backLink)))
+        } else {
 
-          val authorityFiles = request.userAnswers.get(UploadAuthorityPage).getOrElse(Seq.empty).filterNot(_.dutyType == Vat)
+          val previousVATAccountNumber = request.userAnswers.get(AdditionalDefermentNumberPage).getOrElse(dan.accountNumber)
+          val previousVATAccountType = request.userAnswers.get(AdditionalDefermentTypePage).getOrElse(dan.danType)
+          if (dan.accountNumber != previousVATAccountNumber || dan.danType != previousVATAccountType) {
```
Comment (tvaidotas) [inline app/controllers/RepresentativeDanImportVATController.scala:72]: these two are only used in the if condition below, could call a helper method passing user answers that would return a boolean and just call that in the if statement

File: app/controllers/RepresentativeDanImportVATController.scala [hunk 7]
```diff
@@ -59,38 +60,46 @@ class RepresentativeDanImportVATController @Inject()(identify: IdentifierAction,
         backLink
       ))),
       dan => {
-        val previousVATAccountNumber = request.userAnswers.get(AdditionalDefermentNumberPage).getOrElse(dan.accountNumber)
-        val previousVATAccountType = request.userAnswers.get(AdditionalDefermentTypePage).getOrElse(dan.danType)
-        if (dan.accountNumber != previousVATAccountNumber || dan.danType != previousVATAccountType) {
+        val dutyAccountNumber = request.userAnswers.get(DefermentAccountPage)
```
Comment (tvaidotas) [inline app/controllers/RepresentativeDanImportVATController.scala:63]: dutyAccountNumber only used in the if statement, could just have it val dutyAccountNumberExists = request.userAnswers.get(DefermentAccountPage).contains(dan.accountNumber)

**Review by tvaidotas — CHANGES_REQUESTED** [review]
I know it's not your changes per say but would be nice to clean it up a bit

**Review by tvaidotas — APPROVED** [review]

**Pull request merged**